src/shop.c

```diff
@@ -0,0 +1,1368 @@
+            (((u16)sViewportMapObjects[i][X_COORD] << 0x14) + 0xFFF80000) >> 0x10,
+            (u16)sViewportMapObjects[i][Y_COORD] * 16 + 48 - graphicsInfo->height / 2,
+            2);
+        StartSpriteAnim(&gSprites[spriteId], sViewportMapObjects[i][ANIM_NUM]);
+    }
+}
+
+static void BuyMenuCopyTilemapData(void)
+{
+    s16 i;
+    u16 *dst = *gShopTilemapBuffer2;
+    u16 *src = *gShopTilemapBuffer1;
+
+    for (i = 0; i < 0x400; i++)
+    {
+        if (src[i] == 0)
+            continue;
+        dst[i] = src[i] + 0xb3dc;
+    }
+}
+
+static void BuyMenuPrintItemQuantityAndPrice(u8 taskId)
+{
+    s16 *data = gTasks[taskId].data;
+    
+    FillWindowPixelBuffer(3, PIXEL_FILL(1));
+    PrintMoneyAmount(3, 0x36, 0xA, gShopData.itemPrice, TEXT_SPEED_FF);
+    ConvertIntToDecimalStringN(gStringVar1, tItemCount, STR_CONV_MODE_LEADING_ZEROS, 2);
+    StringExpandPlaceholders(gStringVar4, gText_TimesStrVar1);
+    BuyMenuPrint(3, 0, gStringVar4, 2, 0xA, 0, 0, 0, 1);
+}
+
+static void Task_BuyMenu(u8 taskId)
+{
+    s16 *data = gTasks[taskId].data;
+
+    if (!gPaletteFade.active)
+    {
+        s32 itemId = ListMenu_ProcessInput(tListTaskId);
+        ListMenuGetScrollAndRow(tListTaskId, &gShopData.scrollOffset, &gShopData.selectedRow);
+        switch (itemId)
+        {
+        case LIST_NOTHING_CHOSEN:
+            break;
+        case LIST_CANCEL:
+            PlaySE(SE_SELECT);
+            ExitBuyMenu(taskId);
+            break;
+        default:
+            PlaySE(SE_SELECT);
+            tItemId = itemId;
+            ClearWindowTilemap(5);
+            BuyMenuRemoveScrollIndicatorArrows();
+            BuyMenuPrintCursor(tListTaskId, 2);
+            sub_809B10C(1);
+            gShopData.itemPrice = itemid_get_market_price(itemId);
+            if (!IsEnoughMoney(&gSaveBlock1Ptr->money, gShopData.itemPrice))
+            {
+                BuyMenuDisplayMessage(taskId, gText_YouDontHaveMoney, BuyMenuReturnToItemList);
+            }
+            else
+            {
+                CopyItemName(itemId, gStringVar1);
+                BuyMenuDisplayMessage(taskId, gText_Var1CertainlyHowMany, Task_BuyHowManyDialogueInit);
+            }
+            break;
+        }
+    }
+}
+
+static void Task_BuyHowManyDialogueInit(u8 taskId)
+{
+    s16 *data = gTasks[taskId].data;
+    u16 quantityInBag = BagGetQuantityByItemId(tItemId);
+    u16 maxQuantity;
+    
+    BuyMenuQuantityBoxThinBorder(1, 0);
+    ConvertIntToDecimalStringN(gStringVar1, quantityInBag, STR_CONV_MODE_RIGHT_ALIGN, 3);
+    StringExpandPlaceholders(gStringVar4, gText_InBagVar1);
+    BuyMenuPrint(1, 2, gStringVar4, 0, 2, 0, 0, 0, 1);
+    tItemCount = 1;
+    BuyMenuQuantityBoxNormalBorder(3, 0);
+    BuyMenuPrintItemQuantityAndPrice(taskId);
+    ScheduleBgCopyTilemapToVram(0);
+    maxQuantity = GetMoney(&gSaveBlock1Ptr->money) / itemid_get_market_price(tItemId);
+    if (maxQuantity > 99)
+        gShopData.maxQuantity = (u8)99;
```

Comment:
are you sure these casts are necessary lol

Reply:
there's an lsl/lsr 0x18 prior to storing as a halfword - is there a better way?

Reply:
Perhaps a u8 temp variable? I find it weird how Emerald handles this slightly differently (the destination container is u8 instead of u16, and the total cost is stored in memory rather than computed on the fly).

Reply:
tried this as well. Ostensibly it would work, but it removes the lsl/lsr

Reply:
I tried playing with the structure element size along with the temp var, I wasn't able to match it. Could be a combination I missed that works though